titus-server-master/src/main/java/com/netflix/titus/master/jobmanager/service/KubeNotificationProcessor.java

```diff
@@ -135,7 +135,19 @@ public void shutdown() {
     }
 
     private Mono<Void> handlePodUpdatedEvent(PodEvent event, Job job, Task task) {
-        if (event.getPod() == null || event.getPod().getStatus() == null || CollectionsExt.isNullOrEmpty(event.getPod().getStatus().getContainerStatuses())) {
+        // This is basic sanity check. If it fails, we have a major problem with pod state.
+        if (event.getPod() == null || event.getPod().getStatus() == null || event.getPod().getStatus().getPhase() == null) {
+            logger.warn("Pod notification with pod without status or phase set: taskId={}, pod={}", task.getId(), event.getPod());
+            return Mono.empty();
+        }
+
+        PodWrapper podWrapper = new PodWrapper(event.getPod());
+
+        if (!podWrapper.hasContainers()) {
+            if (podWrapper.getPodPhase() == PodPhase.FAILED || podWrapper.getPodPhase() == PodPhase.SUCCEEDED) {
```

Comment:
minor: thoughs on centralizing this logic into `PodWrapper`? e.g.: something like `PodWrapper#isTerminated()`